docs/design/modular_installation.md

```diff
@@ -0,0 +1,63 @@
+# Kuadrant Proposal - Modular Installation
+
+Kuadrant is developing a set of loosely coupled functionalities built directly on top of Kubernetes.
+Kuadrant aims to allow customers to just install, use and understand those functionalities they need.
+Currently, the installation tool of kuadrant, the [kuadrantctl CLI](https://github.com/Kuadrant/kuadrantctl),
+installs all or nothing. For instance, a customer not wanting to rate limit the API,
+should not have to install the rate limit component ([Limitador](https://github.com/Kuadrant/limitador)).
+
+## High Level Goals
+
+* Install only required components. Operate only required components.
+* Expose only the kuadrant API that has available implementation by the with the deployed components
+
+## Installation profiles
+
+The kuadrant installation mechanism should offer modular installation to enable pieces of kuadrant.
+Modular installation options should be feature oriented rather than deployment component oriented.
+Then, it is up to the installation tool to decide what components need to be deployed and how to
+configure it.
+
+Kuadrant features:
+
+* Multiple APIs bundled in a product
+* Service Discovery
+* TLS
+* Authentication (mTLS, API key, OIDC)
+* Rate Limiting (pre-auth, post-auth)
+
+Each feature, or part of it, is eligible to be included or excluded when installing kuadrant.
+
+Some profiles can be defined to group set of commonly required features. Naming the profiles
+allows the customer to easily express wanted installation configuration. Furthermore, profiles
+not only can be used to group a set of features, profiles can be used to define deployment options.
+
+| Name | Description |
+| ---  | --- |
+| **Minimal** | Minimal installation required to run an API without any protection, analytics or API management |
+| **Authenticated** | MTLS, API key and OIDC authentication features |
+| **RateLimit** | Basic rate limit (only pre-auth rate limit) features |
+| **Full** | Full featured kuadrant installation |
+| **Development** | Full featured kuadrant for development purposes. Development deployment mode: debug log level, no cache, tracing |
+
+A kuadrant operator, together with a design of a kuadrant CRD is desired.
+Not only for kuadrant installation, but also for lifecycle management.
+Additionally, the [kuadrantctl](https://github.com/Kuadrant/kuadrantctl) CLI tool can also
+be used to install kuadrant in a profile basis. The CLI tool may also accept, optionally, the kuadrant
+custom resource as input.
```

Comment:
kuadrantctl seems to currently hold the manifest for each kuadrant component (istio, limitador, authorino etc..), but the operator will also need those things. Is the intention here to have kuadrantctl install the operator and configure via the operators CR?

Reply:
I think if we could integrate the ctl and opertor into a single project it could provide some advantages. One thing we would like to do is convert a Open API Spec into k8s kuadrant resources. Being able to do this as part of a CLI or as part of the operator would be valuable for giving end user the option for either

Reply:
As a user myself, my own preference for the ctl would be for it to be as minimal as possible. Converting OAS into K8s resources is a good example of added value. Other than that, I personally would default to maintaining my sets of declarative YAMLs (whose structure I have to learn regardless), and using the k8s CLI, k8s API, rather than learning yet another layer of imperative commands.

Maybe it's only me.

Reply:
For installation purposes, my proposal is that an operator is the first candidate and a CLI is optional. Mainly for covering lifecycle (upgrades) management needs. 

That said, the CLI may add value beyond the installation (and upgrades) features. For instance, as you say, creating k8s objects from OAS or installing the operator itself (useful in k8s without olm), manage kuadrant APIkeys (which are secrets with extras), healthchecks...